### DISCUSSION TITLE
Asignar al documento PUE o PPD según sea el caso

### DISCUSSION THREAD

**Discussion by victorsg60** [AndresRamos/Contpaqi#74] [Q&A]
Buen día
He estado revisando las tablas de documento, concepto de documento, cliente, etc., para saber en que tabla se asigna la forma o método de pago PUE Pago en una sola exhibición y/o PPD Pago en parcialidades o diferido pero por mas que reviso las tablas y hago pruebas NO consigo setear en el documento la forma de pago o método de pago, en específico PUE y/o PPD.
El documento creado se timbra sin ningún problema pero al parecer por defecto le asigna PUE y tengo algunos clientes que su forma de pago o método debe ser PPD
En que tabla se asigna esto al momento de crear el documento?
Por tu atención
Gracias

**Comment by AndresRamos** [comment]
Hola muy buenas tardes, en la documentación puedes encontrar un ejemplo de cómo [Crear Un Documento](https://github.com/AndresRamos/ARSoftware.Contpaqi.Comercial/wiki/Documentos#crear-documento)

```csharp
public class CrearFactura
{
    private readonly IDocumentoService _documentoService;
    private readonly IMovimientoService _movimientoService;

    public CrearFactura(IDocumentoService documentoService, IMovimientoService movimientoService)
    {
        _documentoService = documentoService;
        _movimientoService = movimientoService;
    }

    public int Crear()
    {
        tLlaveDoc siguienteFolio = _documentoService.BuscarSiguienteSerieYFolio("PRUEBAFACTURA");

        var documento = new Documento
        {
            Concepto = new ConceptoDocumento { Codigo = "PRUEBAFACTURA" },
            Serie = siguienteFolio.aSerie,
            Folio = (int)siguienteFolio.aFolio,
            Fecha = DateTime.Today,
            Cliente = new ClienteProveedor { Codigo = "PRUEBA" },
            Referencia = "Referencia",
            FormaPago = FormaPagoEnum._03, // Asigna la forma de pago
            MetodoPago = MetodoPagoEnum.PPD, // Asigna el metodo de pago
            Observaciones = "Observaciones",
            Moneda = MonedaEnum.PesoMexicano.ToMoneda(),
            TipoCambio = 1,
            Agente = new Agente { Codigo = "PRUEBA" }
        };

        documento.DatosExtra.Add(nameof(admDocumentos.CTEXTOEXTRA1), "Texto Extra 1");
        documento.DatosExtra.Add(nameof(admDocumentos.CTEXTOEXTRA2), "Texto Extra 2");

        var movimiento = new Movimiento
        {
            Producto = new Producto { Codigo = "PRUEBA" },
            Almacen = new Almacen { Codigo = "1" },
            Unidades = 1,
            Precio = 100,
            Referencia = "Referencia",
            Observaciones = "Observaciones"
        };

        movimiento.DatosExtra.Add(nameof(admMovimientos.CTEXTOEXTRA1), "Texto Extra 1");

        documento.Movimientos.Add(movimiento);

        int nuevoDocumentoId = _documentoService.Crear(documento);

        foreach (Movimiento mov in documento.Movimientos) _movimientoService.Crear(nuevoDocumentoId, mov);

        return nuevoDocumentoId;
    }
}
```

**Comment by victorsg60** [comment]
Buenas tardes
Creo que tu respuesta se refiere a la versión 5 que recientemente publicaste pero estoy trabajando con la versión 4.4.0 que por cuestiones de mi desarrollo que esta en vb.net y .Net version 4.6.2 no la puedo instalar.
En tu código que me hiciste el favor de enviarme, en la creación de factura (Documento) haces uso de Enumeraciones para asignar FormaPago y MetodoPago para crear el documento.
Desafortunadamente revisando tu codigo no veo que campo de que tabla o tablas estas seteando para modificarlos
Existen tres CMETODOPAG, en Documentos, Conceptos de Documentos y Clientes-Proveedores
El asignar forma de pago en Efectivo, Transferencia, Tarjeta, etc para el documento veo que se asigna desde el Cliente que previamente se le asigno al campo CMETODOPAG como forma de pago predeterminada.
Al crear el documento se le asigna al campo CMETODOPAG  que se utiliza en el documento como forma de pago
En pocas palabras para no enrrollarme tanto mi pregunta sería, que campo "CMETODOPAG" o "???????" y de que tabla se tiene que setear para PUE o para PPD.
Por tu atención, muchísimas gracias
Estoy utilizando AdminpaqSDK porque tengo instalado solamente Factura Electrónica

**Comment by AndresRamos** [comment]
Los campos son:
-  **`CMETODOPAG`** - Campo para asignar la forma de pago
- **`CCANTPARCI`** - Campo para asignar el "método de pago

Te comparto un ejemplo de cómo actualizar el método y forma de pago de un documento utilizando `DocumentoService`:

```csharp
var datosDocumento = new Dictionary<string, string>(documento.DatosExtra);

datosDocumento.TryAdd(nameof(admDocumentos.CMETODOPAG), documento.FormaPago.Value);

datosDocumento.TryAdd(nameof(admDocumentos.CCANTPARCI), MetodoPagoHelper.ConvertToSdkValue(documento.MetodoPago).ToString());

_documentoService.Actualizar(documento.Id, datosDocumento);
```

El campo `CCANTPARCI` es de tipo `int` por lo que no puedes asignar el valor como "PUE" o "PPD", en la libreria te proporciono un convertidor para transformar la enumeración en un valor apropiado para el SDK.

Te comparto el código de `MetodoPagoHelper.ConvertToSdkValue(MetodoPagoEnum metodo);`

```csharp
public static int ConvertToSdkValue(MetodoPagoEnum metodo)
{
    if (metodo == MetodoPagoEnum.PUE) return 1;
    if (metodo == MetodoPagoEnum.PPD) return 2;
    throw new ArgumentException("El metodo no es valido.", nameof(metodo));
}
```

Como puedes ver, si es `PUE` asignas el valor `1` y si es `PPD` asignas el valor `2`.
